// @una-ui/nuxt/dist/runtime/composables/useUnaSettings.js
import { useAppConfig, watch } from "#imports";
import { useStorage } from "@vueuse/core";
import { defu } from "defu";
import { useUnaThemes } from "./useUnaThemes.js";
export function useUnaSettings() {
  const { una } = useAppConfig();
  const { getPrimaryColors, getGrayColors } = useUnaThemes();
  const defaultSettings = {
    primaryColors: una.primary ? getPrimaryColors(una.primary) : {},
    grayColors: una.gray ? getGrayColors(una.gray) : {},
    primary: una.primary,
    gray: una.gray,
    radius: una.radius,
    fontSize: una.fontSize,
    theme: una.theme
  };
  const settings = useStorage("una-settings", defaultSettings, void 0, {
    mergeDefaults: defu
  });
  watch(
    () => [settings.value.primary, settings.value.gray],
    ([primary, gray]) => {
      settings.value.primaryColors = primary ? getPrimaryColors(primary) : {};
      settings.value.grayColors = gray ? getGrayColors(gray) : {};
    },
    { immediate: true }
  );
  function reset() {
    if (una.theme) {
      settings.value.theme = una.theme;
      settings.value.primary = false;
      settings.value.gray = false;
      return;
    }
    settings.value.primary = defaultSettings.primary;
    settings.value.gray = defaultSettings.gray;
    settings.value.fontSize = defaultSettings.fontSize;
    settings.value.radius = defaultSettings.radius;
  }
  return {
    defaultSettings,
    settings,
    reset
  };
}
