// @una-ui/nuxt/dist/runtime/utils/index.js
import { useId } from "vue";
export * from "./cn.js";
export function rgbToHex(r, g, b) {
  return `#${[r, g, b].map((x) => {
    const hex = x.toString(16);
    return hex.length === 1 ? `0${hex}` : hex;
  }).join("")}`;
}
export function hexToRgb(hex) {
  const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
  if (!result)
    throw new Error(`Invalid hex color: ${hex}`);
  return [
    Number.parseInt(result[1], 16),
    Number.parseInt(result[2], 16),
    Number.parseInt(result[3], 16)
  ];
}
export function randomId(prefix) {
  return `una-${prefix}-${useId()}`;
}
export function omitProps(obj, propsToOmit) {
  const newObj = { ...obj };
  propsToOmit.forEach((prop) => delete newObj[prop]);
  return newObj;
}
export function pickProps(obj, propsToPick) {
  const newObj = {};
  propsToPick.forEach((prop) => newObj[prop] = obj[prop]);
  return newObj;
}
export function getPriority(...args) {
  for (const arg of args) {
    if (arg !== void 0)
      return arg;
  }
  return void 0;
}
export function isEqualObject(...objects) {
  return objects.every((obj, i) => {
    if (i === 0)
      return true;
    return JSON.stringify(obj) === JSON.stringify(objects[i - 1]);
  });
}
export function looseToNumber(val) {
  const n = Number.parseFloat(val);
  return Number.isNaN(n) ? val : n;
}
