// @una-ui/nuxt/dist/runtime/components/slots/FormFieldDefault.js
import { cloneVNode, computed, defineComponent } from "vue";
export default defineComponent({
  props: {
    "status": {
      type: String,
      validator: (v) => ["success", "error", "warning", "info"].includes(v),
      required: false
    },
    "name": {
      type: String,
      required: false
    },
    "modelValue": {
      type: [String, Number, Boolean, Object, Array, null, void 0],
      required: false
    },
    "id": {
      type: String,
      required: false
    },
    "aria-describedby": {
      type: String,
      required: false
    },
    "aria-invalid": {
      type: Boolean,
      required: false
    }
  },
  // Add the emits option
  emits: ["blur", "input", "change", "update:modelValue"],
  setup(props, { slots, attrs, emit }) {
    const children = computed(() => slots.default?.()[0].children);
    if (!children.value)
      return () => null;
    const clones = computed(() => children.value?.map((node) => {
      const mergeProps = {
        ...props,
        ...node.props,
        ...attrs,
        "onBlur": (event) => {
          emit("blur", event);
          node.props?.onBlur?.(event);
        },
        "onUpdate:modelValue": (value) => {
          emit("update:modelValue", value);
          node.props?.["onUpdate:modelValue"]?.(value);
        }
      };
      return cloneVNode(node, mergeProps);
    }));
    return () => clones.value;
  }
});
