// @una-ui/nuxt/dist/module.mjs
import { defineNuxtModule, createResolver, installModule, addComponentsDir, addImportsDir, addPlugin } from '@nuxt/kit';
import extendUnocssOptions from './una.config.mjs';
import '@iconify/utils/lib/loader/external-pkg';
import '@una-ui/extractor-vue-script';
import '@una-ui/preset';
import '@una-ui/preset/prefixes';
import 'unocss';
import 'unocss-preset-animations';

const name = "@una-ui/nuxt";
const version = "1.0.0-alpha.12";

const module = defineNuxtModule({
  meta: {
    name,
    configKey: "una",
    version,
    compatibility: {
      nuxt: ">=3.0.0"
    }
  },
  defaults: {
    prefix: "N",
    themeable: true,
    global: true,
    dev: false
  },
  async setup(options, nuxt) {
    const { resolve } = createResolver(import.meta.url);
    nuxt.options.css.unshift(
      import.meta.resolve("@unocss/reset/tailwind.css"),
      import.meta.resolve("@una-ui/preset/una.css")
    );
    nuxt.options.alias["#una"] = resolve("./runtime");
    nuxt.options.appConfig.una = {
      ...{
        primary: "yellow",
        gray: "stone",
        radius: 0.625,
        fontSize: 16,
        theme: null,
        themes: []
      },
      ...nuxt.options.appConfig.una || {}
    };
    nuxt.options.app.rootAttrs = nuxt.options.app.rootAttrs || {};
    nuxt.options.app.rootAttrs.class = [nuxt.options.app.rootAttrs.class, "isolate"].filter(Boolean).join(" ");
    const runtimeDir = resolve("./runtime");
    nuxt.options.build.transpile.push(runtimeDir);
    await installModule(import.meta.resolve("@unocss/nuxt"), extendUnocssOptions(nuxt.options.unocss));
    await installModule(import.meta.resolve("@nuxtjs/color-mode"), {
      classSuffix: "",
      disableTransition: true
    });
    await installModule(import.meta.resolve("@vueuse/nuxt"));
    await installModule(import.meta.resolve("reka-ui/nuxt"), {
      prefix: options.prefix
    });
    await installModule(import.meta.resolve("@vee-validate/nuxt"), {
      componentNames: {
        Form: `${options.prefix}Form`,
        // Field: `${options.prefix}FormField`,
        FieldArray: `${options.prefix}FormFieldArray`,
        ErrorMessage: `${options.prefix}FormErrorMessage`
      }
    });
    addComponentsDir({
      path: resolve("./runtime/components"),
      prefix: options.prefix,
      pathPrefix: false,
      priority: 10
    });
    addImportsDir(resolve(runtimeDir, "composables"));
    if (options.themeable) {
      addPlugin(resolve(runtimeDir, "plugins", "theme.client"));
      addPlugin(resolve(runtimeDir, "plugins", "theme.server"));
    }
    addImportsDir(resolve(runtimeDir, "utils", "cn"));
  }
});

export { module as default };
