// @una-ui/nuxt/dist/una.config.mjs
import { createExternalPackageIconLoader } from '@iconify/utils/lib/loader/external-pkg';
import extratorUna from '@una-ui/extractor-vue-script';
import presetUna from '@una-ui/preset';
import prefixes from '@una-ui/preset/prefixes';
import { transformerDirectives, transformerVariantGroup, presetWind3, presetAttributify, presetIcons, mergeConfigs } from 'unocss';
import { presetAnimations } from 'unocss-preset-animations';

const unaConfig = {
  presets: [
    presetWind3(),
    presetAttributify(),
    presetIcons({
      extraProperties: {
        "display": "inline-block",
        "height": "1em",
        "width": "1em",
        "vertical-align": "middle"
      },
      collections: {
        ...createExternalPackageIconLoader("@iconify-json/lucide", true),
        ...createExternalPackageIconLoader("@iconify-json/radix-icons", true),
        ...createExternalPackageIconLoader("@iconify-json/tabler", true)
      }
    }),
    presetUna(),
    presetAnimations()
  ],
  extractors: [
    extratorUna({
      prefixes
    })
  ],
  transformers: [
    transformerDirectives(),
    transformerVariantGroup()
  ]
};
function defaultConfig(...configs) {
  return mergeConfigs([
    unaConfig,
    ...configs
  ]);
}
function extendUnocssOptions(options = {}) {
  return mergeConfigs([
    unaConfig,
    options
  ]);
}

export { extendUnocssOptions as default, defaultConfig, unaConfig };
